### ISSUE TITLE
Error message about pointer conversion instructs me to write a bug

### ISSUE THREAD

**Issue by notnullnotvoid** [llvm/llvm-project#58958]
![image](https://user-images.githubusercontent.com/37276056/201498472-b6d7e4cc-93d4-4f7a-863d-f2d30e4c68ae.png)

When passing a variable of type `const char *` to a function accepting `void *`, I get an error message like the following:
```console
bob.cpp:184:25: error: no matching function for call to 'MeowHash'
    meow_u128 hash128 = MeowHash(MeowDefaultSeed, strlen(path), path);
                        ^~~~~~~~
./meow_hash_x64_aesni.h:259:1: note: candidate function not viable: no known conversion from 'const char *' to 'void *' for 3rd argument; take the address of the argument with &
```
Presumably because of const-ness. This error message is correct, except for this instruction at the end: `take the address of the argument with &`. In this case the correct thing to is to explicitly cast to `void *`, and if I assumed the compiler knew what it was talking about and followed its instructions, I would write a potentially easy-to-miss pointer bug.

You could argue a programmer should have the dilligence to double-check and not trust the compiler in a case like this, and in principle I agree, but the type of inexperienced programmer who is most likely to find these hints helpful, and most likely to follow them, is also the type most likely to miss a bug like this, so I think this is likely to do harm in practice.

As far as I know there's no way for the compiler to tell what the correct solution in this case is, so the message should probably either not be printed in a case like this with incompatible pointer types, or rephrased to make it clear this is only a suggestion of what *might* be the solution.

**Comment by tbaederr** [comment]
Minimal reproducer: https://godbolt.org/z/r1vrjYs67

**Comment by krsch** [comment]
I want to try fixing this problem, but I need some advice as it is my first issue in LLVM. Do I understand correctly that:
1. This is only a bug if To type is `void *` and From type is a const pointer?
2. The right place to fix it is `lib/Sema/SemaFixItUtils.cpp:137`?
3. I should add a test to `lib/Sema/SemaFixItUtils.cpp`? Or maybe create a new file in the same folder?

**Comment by dwblaikie** [comment]
Actually, now that I think about it - I'm not sure this diagnostic is entirely incorrect. It would be compatible to take the address of the variable and pass it to the void* parameter. So maybe that is the right suggestion here? (admittedly, maybe buggy in some cases - but suggestions are just that, not guarantees)

But, yeah, maybe it's likely enough to be confusing/point the user in the wrong direction, so if the argument is a pointer, maybe that lowers the quality of the "take the address of this" suggestion to the point where we shouldn't make it.


> I want to try fixing this problem, but I need some advice as it is my first issue in LLVM. Do I understand correctly that:
> 
> 1. This is only a bug if To type is `void *` and From type is a const pointer?

Looks like it - at least from my cursory poking around/testing nearby cases.

> 2. The right place to fix it is `lib/Sema/SemaFixItUtils.cpp:137`?

At a glance that looks like the ballpark - maybe the bug is inside `CanConvert`?

(though the diagnostic didn't have a fixit on the note so I'm not 100% sure this is the right place - maybe commenting out this code to check if it has an effect on the diagnostic would help be sure you're in the right place)

> 3. I should add a test to `test/FixIt/fixit-function-call.cpp`? Or maybe create a new file in the same folder?

Yeah, looks like the right place to me.

**Comment by krsch** [comment]
I have submitted my fix to https://reviews.llvm.org/D138426
I don't know any example when I would want to cast `T const * *` to `void*` and pass `T const*` instead as a bug. Much more likely I would forget to do `const_cast` when passing a pointer to C function that would give it back to me in a callback. This use case is very common.

**Comment by AaronBallman** [comment]
Fixed by 2cea4c239570c37f46ad0003b3d41d9473aca60f
